### PR TITLE
ARROW-6926: [Python] Support __sizeof__ protocol for Python objects

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://issues.apache.org/jira/browse/ARROW-6926

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -751,6 +751,9 @@ cdef class Array(_PandasConvertible):
                 size += buf.size
         return size
 
+    def __sizeof__(self):
+        return object.__sizeof__(self) + self.nbytes
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: How about `super(Array, self).__sizeof__() + self.nbytes`?

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -625,6 +629,12 @@ cdef class RecordBatch(_PandasConvertible):
             size += self.column(i).nbytes
         return size
 
+    def __sizeof__(self):
+        return (
+            object.__sizeof__(self) + self.nbytes +
+            sys.getsizeof(self.schema.metadata)
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: I don't think it makes sense to add the schema metadata size, since we are not adding the schema size.

**Review by pitrou — COMMENTED** [review]
Thank you. Just a couple comments.

File: python/pyarrow/table.pxi (hunk 2 above)
Comment (xhochy) [reply python/pyarrow/table.pxi]: Shouldn't we add the schema size? At least in some cases the schema can become quite big.

Comment (pitrou) [reply python/pyarrow/table.pxi]: It's a bit of a fine line, but generally `__sizeof__` shouldn't include the size of objects which are reachable individually. For example:
```python
>>> a = np.zeros(0)                                                                       
>>> sys.getsizeof(a)                                                                      
96
>>> sys.getsizeof(a.dtype)                                                                
96
```

Comment (xhochy) [reply python/pyarrow/table.pxi]: Ah, I had the misconception that `sys.getsizeof` would be recursive. Then we also shouldn't include the metadata size here, too.

Comment (jorisvandenbossche) [reply python/pyarrow/table.pxi]: > It's a bit of a fine line, but generally __sizeof__ shouldn't include the size of objects which are reachable individually

You could say they same about the `buffers` then ... as those are also reachable individually (just to illustrate the thin line ;) I don't want to argue for not including those). 

It's true you can access the schema / schema metadata through a python API, but whether you access this or not, the actual schema (metadata) is stored anyway in the Table (C++) object, so I don't find it necessarily clear on which side of the line it falls.

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the comments!

File: python/pyarrow/table.pxi (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/table.pxi]: Yes, but the `buffers` are not really useful in isolation, while the schema can definitely be applied to multiple tables or batches (and often will). So I'd argue that not including the schema is the most useful here.

Comment (jorisvandenbossche) [reply python/pyarrow/table.pxi]: OK! What about implementing a `__sizeof__` for schema that includes the metadata? Or can metadata be shared across different schema's as well?

Comment (pitrou) [reply python/pyarrow/table.pxi]: > What about implementing a __sizeof__ for schema that includes the metadata?

That sounds fine to me. The metadata can be shared at the C++ level, but I'm not sure that's common.

**Comment by jorisvandenbossche** [comment]
Added a `Schema.__sizeof__`. The problem is that if we actually want to measure the size of the metadata (which can get big), we need to inspect the dict (as a default getsizeof with dict does not measure that).

Now, in principle, fields can also have metadata that is not in the schema metadata ...

**Review by pitrou — APPROVED** [review]
LGTM, thank you!

**Pull request closed without merging**